crates/telio-traversal/src/endpoint_state.rs

```diff
@@ -0,0 +1,119 @@
+use std::fmt;
+use telio_utils::telio_log_warn;
+
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
+pub enum EndpointState {
+    #[default]
+    Disconnected,
+    EndpointGathering,
+    Ping,
+    Published,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Event {
+    SendCallMeMaybeRequest,
+    ReceiveCallMeMaybeResponse,
+    Publish,
+    Timeout,
+    EndpointGone,
+}
+
+#[derive(Default, Debug)]
+pub struct EndpointStateMachine {
+    state: EndpointState,
+    last_event: Option<Event>,
```

Comment:
I think additional states will further increase complexity.

Yes, transitions are handled a bit different in `sm` crate, I think underneath they use dyn traits, consuming the state machine on each transition. This makes the build fail if an invalid transition is attempted. On our case we'll only get a warning in runtime. Do you suggest we should handling it differenttly?

Reply:
`sm` crate seemed to do a really good job if it didn't allow invalid transitions(i didn't know that) :).

My point about `last_event` was that it was used together with the current state to decide about things. This introduces complexity but worse I think it allows for bugs to appear more easily.

As an improvement I suggest to remove `last_event` and represent each unique state separately. There shouldn't be that much of a rework, basically convert two variables into a unique enum.

For example `Disconnected` state is at least issued from these events that are then stored together: `EndpointGone`, `Timeout`. I suggest:
```
enum State {
    Disconnected(DisconnectedReason),
}

enum DisconnectedReason {
    Timeout,
    EndpointGone
}
```

Does that sound reasonable?

Reply:
> Does that sound reasonable?

actually more than reasonable, it sounds better 😅

Reply:
🤔  instead of using `last_state` do evaluate the transition we would use `DisconnectedReason` enum, in the end will be the same logic.

Reply:
And actually `last_state` is giving information that otherwise we would have to implement using the <Disconnected, Ping, EndpointGathering>Reason enums